### ISSUE TITLE
Debug build CI job crashes on `stats._unuran` threading test

### ISSUE THREAD

**Issue by rgommers** [scipy/scipy#14968]
Observed in gh-14962 (unrelated PR). Details from the [CI log](https://github.com/scipy/scipy/runs/4078367919?check_suite_focus=true):

```
sx...................................................................... [ 99%]
......................[gw0] node down: Not properly terminated
F
replacing crashed worker gw0
........................................................ss...ssssssssss. [ 99%]
.s...........................sssssssssssssssssssssssssssssssssssssssssss [ 99%]
ssssssssssssssssys:1: ResourceWarning: unclosed file <_io.TextIOWrapper name=0 mode='r' encoding='UTF-8'>
sys:1: ResourceWarning: unclosed file <_io.TextIOWrapper name=0 mode='r' encoding='UTF-8'>

=================================== FAILURES ===================================
_ build/testenv/lib/python3.8/site-packages/scipy/stats/tests/test_sampling.py _
[gw0] linux -- Python 3.8.0 /usr/bin/python3.8-dbg
worker 'gw0' crashed while running 'build/testenv/lib/python3.8/site-packages/scipy/stats/tests/test_sampling.py::test_threading_behaviour'
```

Likely a bug (threading behavior is something that could indeed cause this segfault), but it's not happening on other PRs. Let's just see if it happens again.

This test came in in commit 402ac3535b8c (main PR that added UNU.RAN, Cc @tirthasheshpatel).

**Comment by tirthasheshpatel** [comment]
> Likely a bug (threading behavior is something that could indeed cause this segfault), but it's not happening on other PRs. Let's just see if it happens again.

I am having a hard time replicating this locally. I fired up 1000 threads but still wasn't able to produce the failure :( I will review the code to check if I can spot something right away.

**Comment by rgommers** [comment]
>  I fired up 1000 threads but still wasn't able to produce the failure 

On a debug build of Python? It could be a build/packaging issue of the interpreter we're using there perhaps. I'd say don't worry too much until it happens again. A code scan would be good, but it's not urgent right now.

**Comment by tirthasheshpatel** [comment]
> On a debug build of Python?

Yes. I used Python3.9-dbg.

**Comment by rgommers** [comment]
This happened once more ([log](https://github.com/scipy/scipy/runs/4675703557?check_suite_focus=true)), again on the `Python38-dbg` job. Output:

```
..sssssssssssssssssssssssssssssssssssssssss [ 99%]
ssssssssssssssssssssssssssssssssssssssssssssssssssssssssssys:1: ResourceWarning: unclosed file <_io.TextIOWrapper name=0 mode='r' encoding='UTF-8'>
sys:1: ResourceWarning: unclosed file <_io.TextIOWrapper name=0 mode='r' encoding='UTF-8'>

=================================== FAILURES ===================================
_ build/testenv/lib/python3.8/site-packages/scipy/stats/tests/test_sampling.py _
[gw0] linux -- Python 3.8.0 /usr/bin/python3.8-dbg
worker 'gw0' crashed while running 'build/testenv/lib/python3.8/site-packages/scipy/stats/tests/test_sampling.py::test_threading_behaviour'
...
=========================== short test summary info ============================
FAILED scipy/stats/tests/test_sampling.py::test_threading_behaviour
...
= 1 failed, 35655 passed, 2599 skipped, 114 xfailed, 7 xpassed in 662.35s (0:11:02) =
```
The crash seems to occur very infrequently.

**Comment by tirthasheshpatel** [comment]
That looks bad. I will try to reproduce and find the root cause.

**Comment by tirthasheshpatel** [comment]
Thankfully, I was able to reproduce:

```python
import threading
from scipy.stats.sampling import TransformedDensityRejection

def test_threading_behaviour():
    # Test if the API is thread-safe.
    # This verifies if the lock mechanism and the use of `PyErr_Occurred`
    # is correct.
    errors = {"err1": None, "err2": None}

    class Distribution:
        def __init__(self, pdf_msg):
            self.pdf_msg = pdf_msg

        def pdf(self, x):
            if 49.9 < x < 50.0:
                raise ValueError(self.pdf_msg)
            return x

        def dpdf(self, x):
            return 1

    def func1():
        dist = Distribution('foo')
        rng = TransformedDensityRejection(dist, domain=(10, 100),
                                          random_state=12)
        try:
            rng.rvs(100000)
        except ValueError as e:
            errors['err1'] = e.args[0]

    def func2():
        dist = Distribution('bar')
        rng = TransformedDensityRejection(dist, domain=(10, 100),
                                          random_state=2)
        try:
            rng.rvs(100000)
        except ValueError as e:
            errors['err2'] = e.args[0]

    t1 = threading.Thread(target=func1)
    t2 = threading.Thread(target=func2)

    t1.start()
    t2.start()

    t1.join()
    t2.join()

    assert errors['err1'] == 'foo'
    assert errors['err2'] == 'bar'

if __name__ == '__main__':
    for _ in range(1000):
        test_threading_behaviour()
```

Error:

![image](https://user-images.githubusercontent.com/43181252/147853849-7facb220-95c4-4e6c-8903-c05b258c46e4.png)

The error occurred because I initiated an attribute lookup in the `finally` block in the `_rvs_cont` method in the base class `Method` once the Python error variable was set and, hence, the object destroyed:

![image](https://user-images.githubusercontent.com/43181252/147853869-e0d65b28-2f3e-47b5-aa96-1a0866fc464f.png)

Notice the `self._callback_wrapper` above. That isn't allowed after the object is garbage collected. I remember facing this issue once while working on the original PR (#14215). I think I forgot to change this at that time. It can be resolved simply by checking if `self._callback_wrapper is not None` at the start of the function and using the boolean value elsewhere. I will submit a patch with the changes.

**Comment by rgommers** [comment]
Great, nice debugging Tirth!

**Comment by tirthasheshpatel** [comment]
Hi @rgommers,

I noticed that I was still sometimes getting attribute lookup errors. I think this is happening because Cython looks up the class attribute to find the `release` method of the `_lock` variable when I call `_lock.release()`. At least, that's what the generated C code says:

![image](https://user-images.githubusercontent.com/43181252/147856543-b526f0d8-8b52-47dc-a080-027c8f6e013b.png)

Although the `_lock` variable is not destroyed, python doesn't allow ANY attribute lookups when a live exception is set. The workaround is to fetch the error before calling `_lock.release()` and restore the error once the lock has been released:

![image](https://user-images.githubusercontent.com/43181252/147856760-a4402e16-da8e-43f4-8546-6f84838a009f.png)

I tried this and reran my experiment 100,000 times. It doesn't give me an attribute lookup error anymore and I am confident this would resolve the issue.

Sorry for not noticing this before #15331 was merged! I will reopen this issue and submit another PR. At some point, it will be better to rewrite the base class in C.
